eclipse-collections/src/main/java/org/eclipse/collections/impl/bimap/mutable/SynchronizedMutableBiMap.java

```diff
@@ -0,0 +1,478 @@
+import org.eclipse.collections.api.block.predicate.Predicate2;
+import org.eclipse.collections.api.block.procedure.Procedure;
+import org.eclipse.collections.api.multimap.set.MutableSetMultimap;
+import org.eclipse.collections.api.ordered.OrderedIterable;
+import org.eclipse.collections.api.partition.set.PartitionMutableSet;
+import org.eclipse.collections.api.set.MutableSet;
+import org.eclipse.collections.api.tuple.Pair;
+import org.eclipse.collections.impl.collection.mutable.SynchronizedMutableCollection;
+import org.eclipse.collections.impl.factory.BiMaps;
+import org.eclipse.collections.impl.list.fixed.ArrayAdapter;
+import org.eclipse.collections.impl.map.AbstractSynchronizedMapIterable;
+import org.eclipse.collections.impl.map.mutable.SynchronizedBiMapSerializationProxy;
+import org.eclipse.collections.impl.map.mutable.SynchronizedMapSerializationProxy;
+import org.eclipse.collections.impl.set.mutable.SynchronizedMutableSet;
+import org.eclipse.collections.impl.utility.LazyIterate;
+
+public class SynchronizedMutableBiMap<K, V> extends AbstractSynchronizedMapIterable<K, V> implements MutableBiMap<K, V>, Serializable
+{
+    private static final long serialVersionUID = 1L;
+
+    protected SynchronizedMutableBiMap(MutableBiMap<K, V> delegate)
+    {
+        super(delegate);
+    }
+
+    protected SynchronizedMutableBiMap(MutableBiMap<K, V> delegate, Object lock)
+    {
+        super(delegate, lock);
+    }
+
+    /**
+     * This method will take a MutableBiMap and wrap it directly in a SynchronizedMutableMap.  It will
+     * take any other non-EC-map and first adapt it will a MapAdapter, and then return a
+     * SynchronizedMutableMap that wraps the adapter.
+     */
+    public static <K, V> SynchronizedMutableBiMap<K, V> of(MutableBiMap<K, V> map)
+    {
+        if (map == null)
+        {
+            throw new IllegalArgumentException("cannot create a SynchronizedMutableBiMap for null");
+        }
+        return new SynchronizedMutableBiMap<K, V>(map);
+    }
+
+    @Override
+    protected MutableBiMap<K, V> getDelegate()
+    {
+        return (MutableBiMap<K, V>) super.getDelegate();
+    }
+
+    public V forcePut(K key, V value)
+    {
+        synchronized (this.lock)
+        {
+            return this.getDelegate().forcePut(key, value);
+        }
+    }
+
+    public MutableBiMap<K, V> asSynchronized()
+    {
+        return this;
+    }
+
+    public MutableBiMap<K, V> asUnmodifiable()
+    {
+        throw new UnsupportedOperationException(this.getClass().getSimpleName() + ".asUnmodifiable() not implemented yet");
+    }
+
+    public MutableBiMap<K, V> clone()
+    {
+        synchronized (this.lock)
+        {
+            return SynchronizedMutableBiMap.of(this.getDelegate().clone());
+        }
+    }
+
+    public MutableBiMap<K, V> tap(Procedure<? super V> procedure)
+    {
+        synchronized (this.lock)
+        {
+            this.forEach(procedure);
+            return this;
```

Comment:
This method doesn't delegate in the way you coded it is because it has to return a MutableBiMap. If we return this.getDelegate().tap(procedure), it will return a non-synchronized copy of the map - isn't that unacceptable? I delegated all methods with the exception of the ones that return "this". For example, withKeyValue(). 

Still, we can change this to:

this.getDelegate().tap(procedure);
return this;

Do you agree?